controllers/controller/devworkspacerouting/util_test.go

```diff
@@ -0,0 +1,236 @@
+		"test-machine-name": {
+			exposedEndpoint,
+			nonExposedEndpoint,
+			discoverableEndpoint,
+		},
+	}
+
+	dwr := &controllerv1alpha1.DevWorkspaceRouting{
+		Spec: controllerv1alpha1.DevWorkspaceRoutingSpec{
+			DevWorkspaceId: workspaceID,
+			RoutingClass:   controllerv1alpha1.DevWorkspaceRoutingBasic,
+			Endpoints:      machineEndpointsMap,
+			PodSelector: map[string]string{
+				constants.DevWorkspaceIDLabel: workspaceID,
+			},
+		},
+	}
+
+	dwr.SetName(name)
+	dwr.SetNamespace(testNamespace)
+
+	Expect(k8sClient.Create(ctx, dwr)).Should(Succeed())
+	return dwr
+}
+
+func getExistingDevWorkspaceRouting(name string) *controllerv1alpha1.DevWorkspaceRouting {
+	By(fmt.Sprintf("Getting existing DevWorkspaceRouting %s", name))
+	dwr := &controllerv1alpha1.DevWorkspaceRouting{}
+	dwrNamespacedName := namespacedName(devWorkspaceRoutingName, testNamespace)
+	Eventually(func() (string, error) {
+		if err := k8sClient.Get(ctx, dwrNamespacedName, dwr); err != nil {
+			return "", err
+		}
+		return dwr.Spec.DevWorkspaceId, nil
+	}, timeout, interval).Should(Not(BeEmpty()), "DevWorkspaceRouting should exist in cluster")
+	return dwr
+}
+
+func getReadyDevWorkspaceRouting(name string) *controllerv1alpha1.DevWorkspaceRouting {
+	dwr := getExistingDevWorkspaceRouting(name)
+	dwrNamespacedName := namespacedName(devWorkspaceRoutingName, testNamespace)
+	Eventually(func() (controllerv1alpha1.DevWorkspaceRoutingPhase, error) {
+		if err := k8sClient.Get(ctx, dwrNamespacedName, dwr); err != nil {
+			return "", err
+		}
+		return controllerv1alpha1.DevWorkspaceRoutingPhase(dwr.Status.Phase), nil
+	}, timeout, interval).Should(Equal(controllerv1alpha1.RoutingReady), "Ready DevWorkspaceRouting should exist in cluster")
+	return dwr
+}
+
+func deleteService(serviceName string, namespace string) {
+	createdService := &corev1.Service{}
+	serviceNamespacedName := namespacedName(serviceName, namespace)
+	Eventually(func() bool {
+		err := k8sClient.Get(ctx, serviceNamespacedName, createdService)
+		return err == nil
+	}, timeout, interval).Should(BeTrue(), "Service should exist in cluster")
+	deleteObject(createdService)
+}
+
+func deleteRoute(endpointName string, namespace string) {
+	createdRoute := routeV1.Route{}
+	routeNamespacedName := namespacedName(common.RouteName(testWorkspaceID, endpointName), namespace)
+	Eventually(func() bool {
+		err := k8sClient.Get(ctx, routeNamespacedName, &createdRoute)
+		return err == nil
+	}, timeout, interval).Should(BeTrue(), "Route should exist in cluster")
+	deleteObject(&createdRoute)
+}
+
+func deleteIngress(endpointName string, namespace string) {
+	createdIngress := networkingv1.Ingress{}
+	ingressNamespacedName := namespacedName(common.RouteName(testWorkspaceID, endpointName), namespace)
+	Eventually(func() bool {
+		err := k8sClient.Get(ctx, ingressNamespacedName, &createdIngress)
+		return err == nil
+	}, timeout, interval).Should(BeTrue(), "Ingress should exist in cluster")
+	deleteObject(&createdIngress)
+}
```

Comment:
What are these required for? Shouldn't services/ingresses/routes/etc. be deleted when the devworkspacerouting is deleted?

Reply:
Will double check :+1:

Reply:
If we don't explicitly delete the routes/services/ingresses, then they will persist and the tests for their owner references will fail, as they are owned by previously created DevWorkspaceRouting objects:

```go
• [FAILED] [0.007 seconds]
DevWorkspaceRouting Controller
/home/aobuchow/git/devworkspace-operator/controllers/controller/devworkspacerouting/devworkspacerouting_controller_test.go:33
  Basic DevWorkspaceRouting Tests
  /home/aobuchow/git/devworkspace-operator/controllers/controller/devworkspacerouting/devworkspacerouting_controller_test.go:34
    OpenShift - DevWorkspaceRouting Objects creation
    /home/aobuchow/git/devworkspace-operator/controllers/controller/devworkspacerouting/devworkspacerouting_controller_test.go:250
      [It] Creates services
      /home/aobuchow/git/devworkspace-operator/controllers/controller/devworkspacerouting/devworkspacerouting_controller_test.go:264

  Begin Captured GinkgoWriter Output >>
    STEP: Getting existing DevWorkspaceRouting test-devworkspacerouting 02/13/23 18:07:46.928
    2023-02-13T18:07:46.929-0500        INFO    controllers.DevWorkspaceRouting Reconciling DevWorkspaceRouting {"Request.Namespace": "devworkspace-test", "Request.Name": "test-devworkspacerouting", "devworkspace_id": "test-id"}
    STEP: Checking single service for all exposed endpoints is created 02/13/23 18:07:46.929
    2023-02-13T18:07:46.932-0500        INFO    controllers.DevWorkspaceRouting Diff:   &v1.Service{
        TypeMeta: {},
        ObjectMeta: v1.ObjectMeta{
                ... // 2 identical fields
                Namespace:                  "devworkspace-test",
                SelfLink:                   "",
    -           UID:                        "",
    +           UID:                        "014f558e-388c-4c7c-872e-8d747383f396",
    -           ResourceVersion:            "",
    +           ResourceVersion:            "291",
                Generation:                 0,
    -           CreationTimestamp:          v1.Time{},
    +           CreationTimestamp:          v1.Time{Time: s"2023-02-13 18:07:44 -0500 EST"},
                DeletionTimestamp:          nil,
                DeletionGracePeriodSeconds: nil,
                Labels:                     {"controller.devfile.io/devworkspace_id": "test-id"},
                Annotations:                nil,
                OwnerReferences: []v1.OwnerReference{
                        {
                                APIVersion: "controller.devfile.io/v1alpha1",
                                Kind:       "DevWorkspaceRouting",
                                Name:       "test-devworkspacerouting",
                                UID: strings.Join({
    -                                   "f81ad575-51f8-4540-830a-7802d698e937",
    +                                   "dbf0612e-84c7-40ae-b642-dc3017df8005",
                                }, ""),
                                Controller:         &true,
                                BlockOwnerDeletion: &true,
                        },
                },
                Finalizers:    nil,
                ClusterName:   "",
    -           ManagedFields: nil,
    +           ManagedFields: []v1.ManagedFieldsEntry{
    +                   {
    +                           Manager:    "devworkspacerouting.test",
    +                           Operation:  "Update",
    +                           APIVersion: "v1",
    +                           Time:       s"2023-02-13 18:07:46 -0500 EST",
    +                           FieldsType: "FieldsV1",
    +                           FieldsV1:   s`{"f:metadata":{"f:labels":{".":{},"f:controller.devfile.io/devwo`...,
    +                   },
    +           },
        },
        Spec: v1.ServiceSpec{
                Ports:                    {{Name: "test-endpoint", Protocol: "TCP", Port: 7777, TargetPort: {IntVal: 7777}, ...}, {Name: "discoverable-endpoint", Protocol: "TCP", Port: 7979, TargetPort: {IntVal: 7979}, ...}},
                Selector:                 {"controller.devfile.io/devworkspace_id": "test-id"},
    -           ClusterIP:                "",
    +           ClusterIP:                "10.0.0.29",
    -           ClusterIPs:               nil,
    +           ClusterIPs:               []string{"10.0.0.29"},
                Type:                     "ClusterIP",
                ExternalIPs:              nil,
    -           SessionAffinity:          "",
    +           SessionAffinity:          "None",
                LoadBalancerIP:           "",
                LoadBalancerSourceRanges: nil,
                ... // 4 identical fields
                SessionAffinityConfig:         nil,
                TopologyKeys:                  nil,
    -           IPFamilies:                    nil,
    +           IPFamilies:                    []v1.IPFamily{"IPv4"},
    -           IPFamilyPolicy:                nil,
    +           IPFamilyPolicy:                &"SingleStack",
                AllocateLoadBalancerNodePorts: nil,
                LoadBalancerClass:             nil,
    -           InternalTrafficPolicy:         nil,
    +           InternalTrafficPolicy:         &"Cluster",
        },
        Status: {},
      }
        {"Request.Namespace": "devworkspace-test", "Request.Name": "test-devworkspacerouting"}
  << End Captured GinkgoWriter Output

  Service should be owned by DevWorkspaceRouting
  Expected
      <[]v1.OwnerReference | len:1, cap:1>: [
          {
              APIVersion: "controller.devfile.io/v1alpha1",
              Kind: "DevWorkspaceRouting",
              Name: "test-devworkspacerouting",
              UID: "dbf0612e-84c7-40ae-b642-dc3017df8005",
              Controller: true,
              BlockOwnerDeletion: true,
          },
      ]
  to contain element matching
      <v1.OwnerReference>: {
          APIVersion: "controller.devfile.io/v1alpha1",
          Kind: "DevWorkspaceRouting",
          Name: "test-devworkspacerouting",
          UID: "f81ad575-51f8-4540-830a-7802d698e937",
          Controller: true,
          BlockOwnerDeletion: true,
      }
  In [It] at: /home/aobuchow/git/devworkspace-operator/controllers/controller/devworkspacerouting/devworkspacerouting_controller_test.go:277
------------------------------


Summarizing 4 Failures:
  [FAIL] DevWorkspaceRouting Controller Basic DevWorkspaceRouting Tests Kubernetes - DevWorkspaceRouting Objects creation [It] Creates services
  /home/aobuchow/git/devworkspace-operator/controllers/controller/devworkspacerouting/devworkspacerouting_controller_test.go:143
  [FAIL] DevWorkspaceRouting Controller Basic DevWorkspaceRouting Tests Kubernetes - DevWorkspaceRouting Objects creation [It] Creates ingress
  /home/aobuchow/git/devworkspace-operator/controllers/controller/devworkspacerouting/devworkspacerouting_controller_test.go:198
  [FAIL] DevWorkspaceRouting Controller Basic DevWorkspaceRouting Tests OpenShift - DevWorkspaceRouting Objects creation [It] Creates route
  /home/aobuchow/git/devworkspace-operator/controllers/controller/devworkspacerouting/devworkspacerouting_controller_test.go:330
  [FAIL] DevWorkspaceRouting Controller Basic DevWorkspaceRouting Tests OpenShift - DevWorkspaceRouting Objects creation [It] Creates services
  /home/aobuchow/git/devworkspace-operator/controllers/controller/devworkspacerouting/devworkspacerouting_controller_test.go:277

```